### PR TITLE
Async Destination: Do not attempt to flush if queue has 0 bytes.

### PR DESCRIPTION
**Pull request by davinchia**
## What
Today we try to flush if a queue has 0 bytes. This can happen if the source is paused for a while. This is both unnecessary and leads to confusing logs.

## How
If a queue has byte size 0, do not attempt to schedule it for work.

## Recommended reading order
All files.

## 🚨 User Impact 🚨
None.

## Pre-merge Actions
*Expand the relevant checklist and delete the others.*

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- Connector version is set to `0.0.1`
    - `Dockerfile` has version `0.0.1`
- Documentation updated
    - Connector's `README.md`
    - Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - `docs/integrations/<source or destination>/<name>.md` including changelog with an entry for the initial version. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - `docs/integrations/README.md`

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

<details><summary><strong>Connector Generator</strong></summary>

- Issue acceptance criteria met
- PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests#pull-request-title-convention)
- If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- Documentation which references the generator is updated as needed

</details>

### CODE REVIEW

File: airbyte-integrations/bases/base-java/src/test/java/io/airbyte/integrations/destination_async/StreamPriorityTest.java [hunk 1]
```diff
@@ -35,14 +35,16 @@ void testOrderByPrioritySize() {
     when(bufferDequeue.getQueueSizeBytes(DESC2)).thenReturn(Optional.of(0L)).thenReturn(Optional.of(1L));
     final DetectStreamToFlush detect = new DetectStreamToFlush(bufferDequeue, null, new AtomicBoolean(false), null);
 
-    assertEquals(List.of(DESC1, DESC2), detect.orderStreamsByPriority(DESCS));
-    assertEquals(List.of(DESC2, DESC1), detect.orderStreamsByPriority(DESCS));
+    // DESC1 has 1 byte, DESC2 has 0 bytes
```
Comment (davinchia) [inline airbyte-integrations/bases/base-java/src/test/java/io/airbyte/integrations/destination_async/StreamPriorityTest.java:38]: I modified the existing tests to catch this. I don't think more tests are needed, however happy to add if needed.

**Comment by octavia-squidington-iii** [comment]
##  source-e2e-test test report (commit [`ea00e64d73`](https://github.com/airbytehq/airbyte/pull/28183/commits/ea00e64d73375afde16486da7a2e8dbbfe95317c)) - ✅

⏲️  Total pipeline duration: 04mn13s 

| Step                                                                       | Result   |
|:---------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-e2e-test/metadata.yaml     | ✅        |
| Connector version semver check                                             | ✅        |
| QA checks                                                                  | ✅        |
| Build connector tar                                                        | ✅        |
| Build source-e2e-test docker image for platform linux/x86_64               | ✅        |
| ./gradlew :airbyte-integrations:connectors:source-e2e-test:integrationTest | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/davinchia_fix-async-error/1689137342/ea00e64d73375afde16486da7a2e8dbbfe95317c/source-e2e-test/2.1.4/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-e2e-test test
```

**Comment by octavia-squidington-iii** [comment]
##  source-alloydb test report (commit [`ea00e64d73`](https://github.com/airbytehq/airbyte/pull/28183/commits/ea00e64d73375afde16486da7a2e8dbbfe95317c)) - ❌

⏲️  Total pipeline duration: 04mn40s 

| Step                                                                      | Result   |
|:--------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-alloydb/metadata.yaml     | ✅        |
| Connector version semver check                                            | ✅        |
| QA checks                                                                 | ✅        |
| Build connector tar                                                       | ✅        |
| Build source-alloydb docker image for platform linux/x86_64               | ✅        |
| ./gradlew :airbyte-integrations:connectors:source-alloydb:integrationTest | ✅        |
| Acceptance tests                                                          | ❌        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/davinchia_fix-async-error/1689137342/ea00e64d73375afde16486da7a2e8dbbfe95317c/source-alloydb/2.0.28/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-alloydb test
```

**Comment by octavia-squidington-iii** [comment]
##  source-mysql test report (commit [`ea00e64d73`](https://github.com/airbytehq/airbyte/pull/28183/commits/ea00e64d73375afde16486da7a2e8dbbfe95317c)) - ❌

⏲️  Total pipeline duration: 15mn52s 

| Step                                                                    | Result   |
|:------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-mysql/metadata.yaml     | ✅        |
| Connector version semver check                                          | ✅        |
| QA checks                                                               | ✅        |
| Build connector tar                                                     | ✅        |
| Build source-mysql docker image for platform linux/x86_64               | ✅        |
| ./gradlew :airbyte-integrations:connectors:source-mysql:integrationTest | ❌        |
| Acceptance tests                                                        | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/davinchia_fix-async-error/1689137342/ea00e64d73375afde16486da7a2e8dbbfe95317c/source-mysql/2.1.0/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-mysql test
```

**Comment by octavia-squidington-iii** [comment]
##  source-mongodb-v2 test report (commit [`ea00e64d73`](https://github.com/airbytehq/airbyte/pull/28183/commits/ea00e64d73375afde16486da7a2e8dbbfe95317c)) - ❌

⏲️  Total pipeline duration: 10mn07s 

| Step                                                                         | Result   |
|:-----------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-mongodb-v2/metadata.yaml     | ✅        |
| Connector version semver check                                               | ✅        |
| QA checks                                                                    | ✅        |
| Build connector tar                                                          | ✅        |
| Build source-mongodb-v2 docker image for platform linux/x86_64               | ✅        |
| ./gradlew :airbyte-integrations:connectors:source-mongodb-v2:integrationTest | ❌        |
| Acceptance tests                                                             | ❌        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/davinchia_fix-async-error/1689137342/ea00e64d73375afde16486da7a2e8dbbfe95317c/source-mongodb-v2/0.2.0/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-mongodb-v2 test
```

**Comment by octavia-squidington-iii** [comment]
##  destination-e2e-test test report (commit [`ea00e64d73`](https://github.com/airbytehq/airbyte/pull/28183/commits/ea00e64d73375afde16486da7a2e8dbbfe95317c)) - ❌

⏲️  Total pipeline duration: 04mn20s 

| Step                                                                            | Result   |
|:--------------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/destination-e2e-test/metadata.yaml     | ✅        |
| Connector version semver check                                                  | ✅        |
| QA checks                                                                       | ✅        |
| Build connector tar                                                             | ✅        |
| Build destination-e2e-test docker image for platform linux/x86_64               | ✅        |
| ./gradlew :airbyte-integrations:connectors:destination-e2e-test:integrationTest | ❌        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/davinchia_fix-async-error/1689137342/ea00e64d73375afde16486da7a2e8dbbfe95317c/destination-e2e-test/0.3.0/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-e2e-test test
```

**Review by benmoriceau — APPROVED** [review]

**Review by evantahler — APPROVED** [review]

**Comment by davinchia** [comment]
- this is not used in any connector except for destination snowflake. 
- I'm reasonably confident this is only better than the current behaviour.

Ran the tests locally:
```
$ SUB_BUILD=CONNECTORS_BASE ./gradlew :airbyte-integrations:bases:base-java:test
...
BUILD SUCCESSFUL in 43s
```

**Review by jbfbell — APPROVED** [review]

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/DetectStreamToFlush.java [hunk 2]
```diff
@@ -203,14 +203,15 @@ long estimateSizeOfRunningWorkers(final StreamDescriptor stream, final long curr
...
    * age of the last record grows, the likelihood of getting any more records from that stream
    * decreases, so by flushing them, we can totally complete that stream). Finally, tertiary sort by
-   * name so the order is deterministic.
+   * name so the order is deterministic. Ignore queues with 0 bytes i.e. no records.
    *
    * @param streams streams to sort.
    * @return streams sorted by priority.
    */
   @VisibleForTesting
   List<StreamDescriptor> orderStreamsByPriority(final Set<StreamDescriptor> streams) {
```
Comment (edgao) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/DetectStreamToFlush.java:212]: nit: rename to `getFlushableStreamsSortedByPriority`? to clarify that it excludes some streams

**Review by edgao — APPROVED** [review]
one tiny nit, nonblocking.

**Comment by octavia-squidington-iii** [comment]
##  destination-redshift test report (commit [`22f71c4bcb`](https://github.com/airbytehq/airbyte/pull/28183/commits/22f71c4bcb6cf2965a1c6614f77dbd70a7cc138f)) - ✅

⏲️  Total pipeline duration: 27mn15s 

| Step                                                                            | Result   |
|:--------------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/destination-redshift/metadata.yaml     | ✅        |
| Connector version semver check                                                  | ✅        |
| QA checks                                                                       | ✅        |
| Build connector tar                                                             | ✅        |
| Build destination-redshift docker image for platform linux/x86_64               | ✅        |
| Build airbyte/normalization-redshift:dev                                        | ✅        |
| ./gradlew :airbyte-integrations:connectors:destination-redshift:integrationTest | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/davinchia_fix-async-error/1689223764/22f71c4bcb6cf2965a1c6614f77dbd70a7cc138f/destination-redshift/0.5.0/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-redshift test
```

**Comment by octavia-squidington-iii** [comment]
##  destination-local-json test report (commit [`22f71c4bcb`](https://github.com/airbytehq/airbyte/pull/28183/commits/22f71c4bcb6cf2965a1c6614f77dbd70a7cc138f)) - ✅

⏲️  Total pipeline duration: 05mn12s 

| Step                                                                              | Result   |
|:----------------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/destination-local-json/metadata.yaml     | ✅        |
| Connector version semver check                                                    | ✅        |
| QA checks                                                                         | ✅        |
| Build connector tar                                                               | ✅        |
| Build destination-local-json docker image for platform linux/x86_64               | ✅        |
| ./gradlew :airbyte-integrations:connectors:destination-local-json:integrationTest | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/davinchia_fix-async-error/1689223764/22f71c4bcb6cf2965a1c6614f77dbd70a7cc138f/destination-local-json/0.2.11/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-local-json test
```

**Comment by davinchia** [comment]
This is going to be replaced by #28258 .

**Pull request closed without merging**